### PR TITLE
Add Connector for Azure OpenAI Chat

### PR DESCRIPTION
**Pull request by JDPrabasha**
# Description

Fixes https://github.com/wso2-enterprise/choreo/issues/19293

Related Pull Requests
- https://github.com/ballerina-platform/openapi-connectors/pull/759
- https://github.com/ballerina-platform/openapi-connectors/pull/760

One line release note: 
- Adds connector for Azure OpenAI Chat

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Ballerina Version: 2201.4.1
* Operating System: Ubuntu 22.04.1 LTS
* Java SDK: openjdk version "1.8.0_362"

# Checklist:

### Changes to OpenAPI definition

- [x] Modified `info` section of the OpenAPI definition file - [Example](https://github.com/ballerina-platform/ballerina-extended-library/discussions/74)

    - `description` - what the connector is about 
    - `x-ballerina-init-description` OpenAPI extension - connector initialization details 
    - `x-ballerina-display` OpenAPI extension - connector name and path to the connector icon 

- [x] Added proper description each API Key, if API Key authentication is defined in OpenAPI definition file 

### Changes to connector module

- [x] Added license header at the top of each .bal file. 
- [x] Added Package.md as per the guide [here](https://github.com/ballerina-platform/ballerina-extended-library/discussions/77)
- [x] Added Module.md for the module of the  connector as per the guide [here](https://github.com/ballerina-platform/ballerina-extended-library/discussions/78)
- [x] Added connector icon to the root of the connector. Icon needs to be a `png` of 200x200 px size with name `icon.png`
- [x] Added Ballerina.toml file with following information. For keywords refer to [guide](https://github.com/ballerina-platform/ballerina-extended-library/discussions/72)

### Verifying connector module

- [x] Compiled the connector successfully with the targeted ballerina version. 
- [ ] Conducted smoke tests on the connector (Recommended when OpenAPI definition is obtain from an unofficial source)

### Security checks
 - [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? 
 - [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: openapi/azure.openai.chat/Ballerina.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[package]
+
+license = ["Apache-2.0"]
+keywords = ["AI/Azure", "AI/OpenAI", "Cost/Paid", "GPT-3.5", "ChatGPT", "Vendor/Microsoft"]
```
Comment (AathmanT) [inline openapi/azure.openai.chat/Ballerina.toml:4]:
```suggestion
keywords = ["AI/Azure", "Cost/Paid", "GPT-3.5", "ChatGPT", "Vendor/Microsoft"]
```

File: openapi/azure.openai.chat/openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    variables:
+      endpoint:
+        default: your-resource-name.openai.azure.com
+security:
+  - bearer:
+      - api.read
+  - apiKey: []
+paths:
+  /deployments/{deployment-id}/chat/completions:
+    post:
```
Comment (AathmanT) [inline openapi/azure.openai.chat/openapi.yaml:27]: Add `x-ballerina-display` to all endpoints. Refer
https://github.com/hevayo/ballerina-openai-client/blob/fixes/client_openapi.yaml#L22-L23

File: openapi/azure.openai.chat/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+target
+generated
+Config.toml
```
Comment (AathmanT) [inline openapi/azure.openai.chat/.gitignore:3]: Shall we remove these

File: openapi/azure.openai.chat/Ballerina.toml (hunk 1 above)
Comment (JDPrabasha) [reply openapi/azure.openai.chat/Ballerina.toml]: Addressed n f713c6745a129c25a6eeda2a853911c8e60245c8

File: openapi/azure.openai.chat/.gitignore (hunk 3 above)
Comment (JDPrabasha) [reply openapi/azure.openai.chat/.gitignore]: Addressed in f713c6745a129c25a6eeda2a853911c8e60245c8

File: openapi/azure.openai.chat/Module.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        serviceUrl = serviceUrl
+    );
+    ```
+
+### Step 3: Invoke connector operation
+1. Now you can use the operations available within the connector.
+
+>**Note:** that they are in the form of remote operations.
+
+    Following is an example on creating a conversation with a OpenAI gpt-35-turbo model:
```
Comment (anupama-pathirage) [inline openapi/azure.openai.chat/Module.md:39]:
```suggestion
    Following is an example of creating a conversation with an OpenAI gpt-35-turbo model:
```

File: openapi/azure.openai.chat/Module.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    final chat:Client chatClient = check new (
+        config = {auth: {apiKey: apiKey}},
+        serviceUrl = serviceUrl
+    );
+
+    chat:Chat_completions_body chatBody = {
+        messages: [{role: "user", content: "What is Ballerina?"}]  
+    };
+
+    chat:Inline_response_200|error unionResult = check chatClient->/deployments/["chat"]/chat/completions.post("2023-03-15-preview", chatBody);
```
Comment (anupama-pathirage) [inline openapi/azure.openai.chat/Module.md:53]: Incorrect code, error cannot be return as check is there.

File: openapi/azure.openai.chat/Module.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
+## Overview
+This is a generated connector from [Azure OpenAI Chat Completions API](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#chat-completions/) OpenAPI specification.
+
+The Azure Azure OpenAI Service REST API Chat Completions Endpoint will create completions for chat messages with the ChatGPT (preview) and GPT-4 (preview) models.
+
+## Prerequisites
+- Create an [Azure](https://azure.microsoft.com/en-us/features/azure-portal/) account
+- Create an [Azure OpenAI resource](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource)
+- Deploy an appropriate model within the resource by referring to [this guide](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource?pivots=web-portal#deploy-a-model)
```
Comment (anupama-pathirage) [inline openapi/azure.openai.chat/Module.md:9]:
```suggestion
- Deploy an appropriate model within the resource by referring to [Deploy a model](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource?pivots=web-portal#deploy-a-model) guide.
```

File: openapi/azure.openai.chat/Module.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+- Obtain tokens by following [this guide](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#authentication)
```
Comment (anupama-pathirage) [inline openapi/azure.openai.chat/Module.md:10]:
```suggestion
- Obtain tokens by following [Azure OpenAI Authentication](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#authentication) guide
```

File: openapi/azure.openai.chat/Module.md (hunk 5 above)
Comment (JDPrabasha) [reply openapi/azure.openai.chat/Module.md]: Addressed in 21e3fcbcb761b6eaa57086c7af39f00940473e6a

File: openapi/azure.openai.chat/Module.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    ```
+
+### Step 3: Invoke connector operation
+1. Now you can use the operations available within the connector.
+
+>**Note:** that they are in the form of remote operations.
+
+    Following is an example on creating a conversation with a OpenAI gpt-35-turbo model:
+
+    ```ballerina
```
[the remark this answers is not in the document]
Comment (JDPrabasha) [reply openapi/azure.openai.chat/Module.md:43]: Addressed in eb85be029893d9a81a22255df84d75e338797609

File: openapi/azure.openai.chat/Module.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    );
+
+    chat:Chat_completions_body chatBody = {
+        messages: [{role: "user", content: "What is Ballerina?"}]  
+    };
+
+    chat:Inline_response_200 chatResult = check chatClient->/deployments/["chat"]/chat/completions.post("2023-03-15-preview", chatBody);
+
+    io:println(chatResult);
+    }
```
Comment (anupama-pathirage) [inline openapi/azure.openai.chat/Module.md:58]: Code within the main is not properly indented.

Comment (JDPrabasha) [reply openapi/azure.openai.chat/Module.md:58]: Addressed in 1e8102ca7d7627e1000cd07c3ac4146226f0b855

File: openapi/azure.openai.chat/original-openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,817 @@
+{
```
Comment (NipunaRanasinghe) [inline openapi/azure.openai.chat/original-openapi.yaml:1]: I'm seeing 2 openAPI files being added as `openapi.yaml` and `original-openapi.yaml`. Why do we need to keep both?

Comment (JDPrabasha) [reply openapi/azure.openai.chat/original-openapi.yaml:1]: We are splitting a spec into 3 to generate 3 separate connectors. As discussed in a meeting regarding this, we decided to keep the original spec in each connector directory (similar to how it is in the Stripe connector). This is mainly done to be able to trace the versions of the original spec when they get updated

Comment (NipunaRanasinghe) [reply openapi/azure.openai.chat/original-openapi.yaml:1]: @JDPrabasha I see.. makes sense

**Pull request closed without merging**